src/ocean/util/container/btree/Implementation.d

```diff
@@ -0,0 +1,1200 @@
+        foreach (i, ref el; dest.elements[old_number..dest.number_of_elements])
+        {
+            el = source.elements[source_start + i];
+        }
+
+        if (!dest.is_leaf)
+        {
+            // TODO: assert (!souce.is_leaf)
+            foreach (i, ref child_node;
+                    dest.child_nodes[old_number..dest.number_of_elements+1])
+            {
+                child_node = source.child_nodes[source_start + i];
+            }
+        }
+
+        source.number_of_elements = source_start;
+    }
+
+    /***************************************************************************
+
+        Removes the element from the node. This removes the element from the node
+        and reorganizes the subtrees.
+
+        Params:
+            node = node to remove from
+            position = position of the element to remove.
+
+        Returns:
+            the removed element.
+
+    ***************************************************************************/
+
+    private static BTreeMapNode.KeyValue popFromNode (BTreeMapNode* node, size_t position)
+    {
+        auto element = node.elements[position];
+
+        // rotate the next neighbour elements
+        for (auto i = position; i < node.number_of_elements-1; i++)
+        {
+            node.elements[i] = node.elements[i+1];
+        }
+
+        if (!node.is_leaf)
+        {
+            for (auto i = position; i < node.number_of_elements; i++)
+            {
+                node.child_nodes[i] = node.child_nodes[i+1];
+            }
+        }
+
+        node.number_of_elements--;
+        return element;
+    }
+
+    /******************************************************************************
+
+        Finds the maxima in the subtree.
+
+        Params:
+            node = root of the (sub)tree
+            index = element index in the node.
+
+        Returns:
+            pointer to the node containing the maximum element
+
+    ******************************************************************************/
+
+    private static BTreeMapNode* findMaximum (BTreeMapNode* node,
+        out size_t index)
+    {
+        if (node.is_leaf)
+        {
+            index = node.number_of_elements - 1;
+            return node;
+        }
+        else
+        {
+            return findMaximum(node.child_nodes[node.number_of_elements], index);
+        }
+    }
+
+    /******************************************************************************
+
+        Visits the tree in the inorder order.
+
+        Params:
+            tree = tree to visit
+            dg = delegate to call for each visited element. Delegate should return
+                 non-zero value if the visiting should be aborted.
+
+        Returns:
+            return value of the last dg call.
+
+    ******************************************************************************/
+
+    package int inorder (int delegate(ref KeyType key, ref ValueType value) dg)
```

Comment:
Might be nice to combine the two `inorder` overloads. A template with a `static if` to call the different delegate types?